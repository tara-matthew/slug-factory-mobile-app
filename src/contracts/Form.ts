import { IInputWithLabel } from "./InputWithLabel";

export interface IFormProps {
    inputs: IInputWithLabel[];
    buttonText: string;
    buttonTo: string;
    sendDataToParent: (data: { [value: string]: string }) => void;
}
