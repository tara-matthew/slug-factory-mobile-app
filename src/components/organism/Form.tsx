import React, { useState } from "react";
import { View } from "react-native";
import BaseButton from "../atom/BaseButton";
import InputGroup from "../molecule/InputGroup";
import {IFormProps} from "../../contracts/Form";

const Form = ({ inputs, buttonText, buttonTo, sendDataToParent }: IFormProps) => {
    const [formValues, setFormValues] = useState(
        inputs.reduce((acc, field) => ({ ...acc, [field.placeholder]: "" }), {}),
    );

    const handleSubmit = async () => {
        sendDataToParent(formValues);
    };

    const handleChange = (name: string, value: string) => {
        setFormValues({ ...formValues, [name]: value });
    };

    return (
        <View className="w-full">
            <InputGroup inputs={ inputs } sendDataToParent={ handleChange }></InputGroup>
            <View className="w-full mt-4">
                <BaseButton title={ buttonText } sendDataToParent={ handleSubmit }></BaseButton>
            </View>
        </View>
    );
};

export default Form;
