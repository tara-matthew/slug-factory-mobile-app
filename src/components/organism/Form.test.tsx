import { render } from "@testing-library/react-native";
import Form from "./Form";
import React from "react";

it("renders correctly", () => {
    const mockSendDataToParent = jest.fn();
    const inputs = [
        {
            title: "First input",
            placeholder: "First placeholder",
            isMultiline: false,
        },
        {
            title: "Second input",
            placeholder: "Second placeholder",
            isMultiline: false,
        },
        {
            title: "Third input",
            placeholder: "Third placeholder",
            isMultiline: false,
        },
    ];

    const buttonText: string = "Click me";
    const { toJSON } = render(<Form inputs={ inputs } buttonText={ buttonText } sendDataToParent={ mockSendDataToParent } />);
    expect(toJSON()).toMatchSnapshot();
});

it("displays the correct inputs", () => {
    const mockSendDataToParent = jest.fn();
    const inputs = [
        {
            title: "Name",
            placeholder: "Enter name",
            isMultiline: false,
        },
        {
            title: "Email",
            placeholder: "Enter email",
            isMultiline: false,
        },
        {
            title: "Password",
            placeholder: "Enter password",
            isMultiline: false,
        },
    ];

    const buttonText: string = "Click me";
    const buttonTo: string = "To link";
    const { getByPlaceholderText, getByText } = render(<Form inputs={ inputs } buttonText={ buttonText } buttonTo={ buttonTo } sendDataToParent={ mockSendDataToParent } />);

    const placeholders = [getByPlaceholderText("Enter name"), getByPlaceholderText("Enter email"), getByPlaceholderText("Enter password")];
    placeholders.forEach(placeholder => expect(placeholder).toBeTruthy());

    const inputTexts = [getByText("Name"), getByText("Email"), getByText("Password")];
    inputTexts.forEach(input => expect(input).toBeTruthy());
});

it.todo("sends data when the button is clicked");
it.todo("populates the fields");
