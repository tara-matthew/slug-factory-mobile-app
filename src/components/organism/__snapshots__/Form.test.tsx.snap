// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  className="w-full"
>
  <View>
    <Text
      className="font-bold"
    />
    <View
      className="bg-black/5 w-full p-5 rounded-2xl mb-7"
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="First placeholder"
        value=""
      />
    </View>
  </View>
  <View>
    <Text
      className="font-bold"
    />
    <View
      className="bg-black/5 w-full p-5 rounded-2xl mb-7"
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Second placeholder"
        value=""
      />
    </View>
  </View>
  <View>
    <Text
      className="font-bold"
    />
    <View
      className="bg-black/5 w-full p-5 rounded-2xl mb-7"
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Third placeholder"
        value=""
      />
    </View>
  </View>
  <View
    className="w-full mt-4"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#d0cadb",
          "opacity": 1,
        }
      }
    >
      <Text
        className="text-center"
      >
        Click me
      </Text>
    </View>
  </View>
</View>
`;
